### PR TITLE
Fix #744 : Highfi : Topic Practice

### LINKED ISSUE #744: Highfi: Topic Practice
TopicPractice needs to be pixel perfect with the mocks.
Portrait Mock: https://xd.adobe.com/view/e8aa4198-3940-47f9-514a-f41cc54457f6-9e9b/screen/f403add1-079a-4202-9975-efe04c076290/TP-Practice-Unchecked-
Landscape Mock: https://xd.adobe.com/view/ee9e607b-dbd6-4372-48dc-b687d32af3da-98af/screen/0fa47d7c-e10d-4097-8773-0efa665a323f/TP-Practice-

Mocks have been created for MDPI Devices.
Make sure that the screen passes the accessibility test also.
For reference you can checkout: https://github.com/oppia/oppia-android/wiki/Working-on-UI

### PR DESCRIPTION
**Pull request by virajpatel2411**
## Explanation 

UI Corrected and made pixel perfect as the mockups.
Issue solved #744.

## Mocks:

Portrait Mock: https://xd.adobe.com/view/e8aa4198-3940-47f9-514a-f41cc54457f6-9e9b/screen/f403add1-079a-4202-9975-efe04c076290/TP-Practice-Unchecked-
Landscape Mock: https://xd.adobe.com/view/ee9e607b-dbd6-4372-48dc-b687d32af3da-98af/screen/0fa47d7c-e10d-4097-8773-0efa665a323f/TP-Practice-

### CODE REVIEW

**Comment by virajpatel2411** [comment]
@rt4914 Can you please review it now?

**Comment by veena14cs** [comment]
Please add mock links as well for reference in this PR.

**Comment by virajpatel2411** [comment]
Portrait Mock: https://xd.adobe.com/view/e8aa4198-3940-47f9-514a-f41cc54457f6-9e9b/screen/f403add1-079a-4202-9975-efe04c076290/TP-Practice-Unchecked-
Landscape Mock: https://xd.adobe.com/view/ee9e607b-dbd6-4372-48dc-b687d32af3da-98af/screen/0fa47d7c-e10d-4097-8773-0efa665a323f/TP-Practice-

File: app/src/main/res/layout-land/topic_practice_footer_view.xml [hunk 1]
```diff
@@ -15,16 +15,15 @@
 
   <androidx.constraintlayout.widget.ConstraintLayout
     android:layout_width="match_parent"
-    android:layout_height="wrap_content"
-    android:paddingStart="96dp"
-    android:paddingEnd="96dp">
+    android:layout_height="wrap_content">
 
     <Button
```
Comment (veena14cs) [inline app/src/main/res/layout-land/topic_practice_footer_view.xml:20]: Check the color of Start button its not white in the mock when it is inactive..

File: app/src/main/res/layout-land/topic_practice_header_view.xml [hunk 2]
```diff
@@ -12,12 +12,10 @@
   <androidx.constraintlayout.widget.ConstraintLayout
     android:layout_width="match_parent"
     android:layout_height="wrap_content"
-    android:layout_marginStart="32dp"
     android:layout_marginTop="32dp"
-    android:layout_marginEnd="24dp"
-    android:layout_marginBottom="8dp"
-    android:paddingStart="48dp"
-    android:paddingEnd="48dp">
+    android:layout_marginLeft="96dp"
```
Comment (veena14cs) [inline app/src/main/res/layout-land/topic_practice_header_view.xml]: Use marginStart and marginEnd to support RTL here and elsewhere.

File: app/src/main/res/layout-land/topic_practice_subtopic.xml [hunk 3]
```diff
@@ -16,8 +16,9 @@
   <androidx.constraintlayout.widget.ConstraintLayout
     android:layout_width="match_parent"
     android:layout_height="wrap_content"
-    android:paddingStart="48dp"
-    android:paddingEnd="48dp"
+    android:layout_marginBottom="16dp"
+    android:layout_marginRight="96dp"
```
Comment (veena14cs) [inline app/src/main/res/layout-land/topic_practice_subtopic.xml]: Recommended to use margin start and margin end.

File: app/src/main/res/layout-land/topic_practice_subtopic.xml [hunk 4]
```diff
@@ -31,17 +32,15 @@
...
       android:checked="@{isChecked}"
       android:fontFamily="sans-serif"
-      android:gravity="top"
-      android:paddingStart="20dp"
-      android:paddingEnd="20dp"
+      android:gravity="center_vertical"
       android:text="@{viewModel.subtopic.title}"
       android:textColor="@color/oppiaPrimaryText"
       android:textSize="16sp"
+      android:drawableLeft="@android:color/transparent"
```
Comment (veena14cs) [inline app/src/main/res/layout-land/topic_practice_subtopic.xml]: Ditto change it to drawableStart.

File: app/src/main/res/layout/topic_practice_header_view.xml [hunk 5]
```diff
@@ -6,13 +6,15 @@
...
+
   <androidx.constraintlayout.widget.ConstraintLayout
     android:layout_width="match_parent"
     android:layout_height="wrap_content"
-    android:layout_marginStart="32dp"
+    android:layout_marginLeft="32dp"
     android:layout_marginTop="32dp"
-    android:layout_marginEnd="24dp"
-    android:layout_marginBottom="8dp">
+    android:layout_marginRight="24dp"
```
Comment (veena14cs) [inline app/src/main/res/layout/topic_practice_header_view.xml]: Ditto.

File: app/src/main/res/layout/topic_practice_subtopic.xml [hunk 6]
```diff
@@ -32,12 +35,10 @@
       android:layout_height="wrap_content"
       android:button="@drawable/selector_checkbox"
       android:checked="@{isChecked}"
+      android:drawableLeft="@android:color/transparent"
```
Comment (veena14cs) [inline app/src/main/res/layout/topic_practice_subtopic.xml]: Ditto.

**Comment by virajpatel2411** [comment]
@veena14cs Have done the required changes

**Comment by virajpatel2411** [comment]
New image of button is as shown

**Comment by nikitamarysolomanpvt** [comment]
@virajpatel2411 Please click on the Resolved conversation  Button after resolving and commenting.

File: app/src/main/res/layout-land/topic_practice_header_view.xml (hunk 2 above)
Comment (virajpatel2411) [reply app/src/main/res/layout-land/topic_practice_header_view.xml]: Changed

File: app/src/main/res/layout-land/topic_practice_subtopic.xml (hunk 4 above)
Comment (virajpatel2411) [reply app/src/main/res/layout-land/topic_practice_subtopic.xml]: Done

File: app/src/main/res/layout/topic_practice_subtopic.xml (hunk 6 above)
Comment (virajpatel2411) [reply app/src/main/res/layout/topic_practice_subtopic.xml]: Done

**Comment by virajpatel2411** [comment]
@nikitamarysolomanpvt  Done. Can you please finally review it and merge if possible?

File: app/src/main/res/layout-land/topic_practice_subtopic.xml (hunk 3 above)
Comment (virajpatel2411) [reply app/src/main/res/layout-land/topic_practice_subtopic.xml]: Done

File: app/src/main/res/layout-land/topic_practice_subtopic.xml [hunk 7]
```diff
@@ -31,17 +32,15 @@
       android:id="@+id/subtopic_check_box"
```
Comment (nikitamarysolomanpvt) [inline app/src/main/res/layout-land/topic_practice_subtopic.xml:32]: It seems failing accessibility test.

**Comment by nikitamarysolomanpvt** [comment]
@mschanteltc accessibility test is failing for the checkbox PTAL.

File: app/src/main/res/layout-land/topic_practice_subtopic.xml (hunk 7 above)
Comment (virajpatel2411) [reply app/src/main/res/layout-land/topic_practice_subtopic.xml:32]: I am not getting what needs to be done?

**Comment by mschanteltc** [comment]
> @mschanteltc accessibility test is failing for the checkbox PTAL.

Is this because it does not meet the minimum touch target size? Or any other accessibility issue?

File: app/src/main/res/layout/topic_practice_fragment.xml [hunk 8]
```diff
@@ -12,7 +12,7 @@
     android:layout_height="match_parent"
     android:clipToPadding="false"
     android:overScrollMode="never"
-    android:paddingBottom="172dp"
+    android:layout_marginTop="12dp"
```
Comment (veena14cs) [inline app/src/main/res/layout/topic_practice_fragment.xml:15]: You can remove this margin top. As you are setting margin top 32 dp in topic_practice_header_view.

File: app/src/main/res/layout/topic_practice_header_view.xml [hunk 9]
```diff
@@ -21,18 +23,20 @@
       android:text="@string/topic_practice_master_these_skills"
       android:textColor="@color/oppiaPrimaryText"
       android:textSize="20sp"
+      android:layout_marginEnd="88dp"
```
Comment (veena14cs) [inline app/src/main/res/layout/topic_practice_header_view.xml:26]: You can give margin end as 32 dp for consistency.

File: app/src/main/res/layout/topic_practice_subtopic.xml [hunk 10]
```diff
@@ -32,12 +35,10 @@
       android:layout_height="wrap_content"
       android:button="@drawable/selector_checkbox"
       android:checked="@{isChecked}"
+      android:drawableStart="@android:color/transparent"
```
Comment (veena14cs) [inline app/src/main/res/layout/topic_practice_subtopic.xml:39]: Give min height as 48dp to pass accessibility test.

File: app/src/main/res/layout-land/topic_practice_subtopic.xml (hunk 7 above)
Comment (veena14cs) [reply app/src/main/res/layout-land/topic_practice_subtopic.xml:32]: Clickable item should be 48dp or larger. In that case we are not supposed to give fixed height, so you can give min height as 48 dp and height as wrap_content. Make changes in both portrait and landscape to pass accessibility test.

**Comment by veena14cs** [comment]
> > @mschanteltc accessibility test is failing for the checkbox PTAL.
> 
> Is this because it does not meet the minimum touch target size? Or any other accessibility issue?

Yes it is because of minimum touch target size. We need to make it 48dp.

**Comment by nikitamarysolomanpvt** [comment]
> > @mschanteltc accessibility test is failing for the checkbox PTAL.
> 
> Is this because it does not meet the minimum touch target size? Or any other accessibility issue?

Yes checkbox size does not meet the minimum touch target size

File: app/src/main/res/layout-land/topic_practice_subtopic.xml (hunk 7 above)
Comment (virajpatel2411) [reply app/src/main/res/layout-land/topic_practice_subtopic.xml:32]: Okay

File: app/src/main/res/layout/topic_practice_fragment.xml (hunk 8 above)
Comment (virajpatel2411) [reply app/src/main/res/layout/topic_practice_fragment.xml:15]: No in header I am setting 24dp only...So adding 12dp to 24dp total makes 32dp. I have done in  that way.

File: app/src/main/res/layout/topic_practice_header_view.xml (hunk 9 above)
Comment (virajpatel2411) [reply app/src/main/res/layout/topic_practice_header_view.xml:26]: Done

File: app/src/main/res/layout/topic_practice_subtopic.xml (hunk 10 above)
Comment (virajpatel2411) [reply app/src/main/res/layout/topic_practice_subtopic.xml:39]: But then the margin between the check boxes is getting increased. @veena14cs

Comment (virajpatel2411) [reply app/src/main/res/layout/topic_practice_subtopic.xml:39]: I am giving the checkbox height and width of 14dp. Will it be Okay by it?

File: app/src/main/res/layout/topic_practice_fragment.xml (hunk 8 above)
Comment (veena14cs) [reply app/src/main/res/layout/topic_practice_fragment.xml:15]: Ok got it. Make it consistent in both portrait and landscape.

File: app/src/main/res/layout/topic_practice_subtopic.xml (hunk 10 above)
Comment (veena14cs) [reply app/src/main/res/layout/topic_practice_subtopic.xml:39]: Don't give fixed width and height. Share screenshot after giving minHeight.

Comment (virajpatel2411) [reply app/src/main/res/layout/topic_practice_subtopic.xml:39]: The margins are getting increased

File: app/src/main/res/layout/topic_practice_fragment.xml (hunk 8 above)
Comment (virajpatel2411) [reply app/src/main/res/layout/topic_practice_fragment.xml:15]: Done same as mockup

File: app/src/main/res/layout/topic_practice_subtopic.xml (hunk 10 above)
Comment (virajpatel2411) [reply app/src/main/res/layout/topic_practice_subtopic.xml:39]: @veena14cs The screenshot when just minHeight of 48dp given and no margin

**Comment by mschanteltc** [comment]
I updated the [mock](https://xd.adobe.com/view/e8aa4198-3940-47f9-514a-f41cc54457f6-9e9b/screen/f403add1-079a-4202-9975-efe04c076290/TP-Practice-Unchecked-/) and increased the checkbox to be 16px in height and width and increase the textbox size (not font) to be at least 22 px in height. These two can be tapped on to check the box. The options are spaced out 26 px apart, so each option has 13px top and bottom padding, meeting the minimum touch target requirement.

**Comment by virajpatel2411** [comment]
@veena14cs @rt4914 @nikitamarysolomanpvt @mschanteltc Please review. I have updated as per new mock

**Comment by mschanteltc** [comment]
> @veena14cs @rt4914 @nikitamarysolomanpvt @mschanteltc Please review. I have updated as per new mock

Spacing LGTM! Can you share a screenshot where a Subtopic's textbox takes up two lines?

**Comment by virajpatel2411** [comment]
> > @veena14cs @rt4914 @nikitamarysolomanpvt @mschanteltc Please review. I have updated as per new mock
> 
> Spacing LGTM! Can you share a screenshot where a Subtopic's textbox takes up two lines?

Imk the text corresponding to checkbox is coming dynamically and at present there is no text longer enough to take two lines

**Comment by nikitamarysolomanpvt** [comment]
> > > @veena14cs @rt4914 @nikitamarysolomanpvt @mschanteltc Please review. I have updated as per new mock
> > 
> > 
> > Spacing LGTM! Can you share a screenshot where a Subtopic's textbox takes up two lines?
> 
> Imk the text corresponding to checkbox is coming dynamically and at present there is no text longer enough to take two lines

Please share the updated screen shots so that it will be easy for @mschanteltc to help you out.
Iam unassigning me.Please assign me back once its done and accessibility test is passed.

**Comment by virajpatel2411** [comment]
New Screenshots: 




Please review the changes. @rt4914 @veena14cs @nikitamarysolomanpvt @mschanteltc

**Comment by veena14cs** [comment]
@virajpatel2411 can you please mention your branch name. Branch name should always contain underscore and not spaces and must be in small case.

**Comment by virajpatel2411** [comment]
> @virajpatel2411 can you please mention your branch name. Branch name should always contain underscore and not spaces and must be in small case.

Due to conflicts with other PR I have deleted the repo before a day. So now the branch name isn't visible.
Will it get merged?
@veena14cs @rt4914 @nikitamarysolomanpvt

**Comment by rt4914** [comment]
> > @virajpatel2411 can you please mention your branch name. Branch name should always contain underscore and not spaces and must be in small case.
> 
> Due to conflicts with other PR I have deleted the repo before a day. So now the branch name isn't visible.
> Will it get merged?
> @veena14cs @rt4914 @nikitamarysolomanpvt

No actually because even we cannot check this branch because that branch does not exist. Unless and until we are sure that the code we are merging is correct, we cannot merge the code. So I  think in this case you will need to rework.

**Comment by rt4914** [comment]
@virajpatel2411 Closing this PR as the branch no longer exists.

**Comment by virajpatel2411** [comment]
> > > @virajpatel2411 can you please mention your branch name. Branch name should always contain underscore and not spaces and must be in small case.
> > 
> > 
> > Due to conflicts with other PR I have deleted the repo before a day. So now the branch name isn't visible.
> > Will it get merged?
> > @veena14cs @rt4914 @nikitamarysolomanpvt
> 
> No actually because even we cannot check this branch because that branch does not exist. Unless and until we are sure that the code we are merging is correct, we cannot merge the code. So I think in this case you will need to rework.

I was telling you to review from last 10 days. You requested changes, made changes to mock, I did everthing, changed every change you told me to make, still you don't approve and keep the pr untouched. I will do the work again but this was not fair from your side. @rt4914

**Comment by rt4914** [comment]
> > > > @virajpatel2411 can you please mention your branch name. Branch name should always contain underscore and not spaces and must be in small case.
> > > 
> > > 
> > > Due to conflicts with other PR I have deleted the repo before a day. So now the branch name isn't visible.
> > > Will it get merged?
> > > @veena14cs @rt4914 @nikitamarysolomanpvt
> > 
> > 
> > No actually because even we cannot check this branch because that branch does not exist. Unless and until we are sure that the code we are merging is correct, we cannot merge the code. So I think in this case you will need to rework.
> 
> I was telling you to review from last 10 days. You requested changes, made changes to mock, I did everything, changed every change you told me to make, still you don't approve and keep the pr untouched. I will do the work again but this was not fair from your side. @rt4914

@virajpatel2411 
If i am seeing things correctly, @veena14cs and @nikitamarysolomanpvt have been reviewing your PR on a regular basis. Unless and until your PR is not perfect we cannot merge that PR even if takes a month long time. So I actually don't thing that there has been delay from our side. 
Also, patience is one of the most important part of the development process. Unless and until you do not have that actually it can become hard to work. I would suggest you to follow these steps whenever you are creating a PR.

To improvise on your PR quality you can follow these steps:

1. Once your PR is ready, go to `Files Changed`.
2. Check if there are any unnecessary files changes which you have not introduced but are still showing because of android studio, then revert those changes.
3. Now go through each file in `Files Changed` and open that same file in Android Studio and do following steps:
 - `Control + Alt + L` - formatting shortcut.
 - Try to solve all possible warnings given by android studio.
4. Once all these changes are done, run the test cases which you have written in this PR, if any.

Once all this is done, commit again if there are any changes and then assign it to the reviewers.

**Comment by virajpatel2411** [comment]
@rt4914 The thing is from last few days...the code was perfect. I had made all the changes requested. I am patient and not blaming you. I am new to this open source, don't know much so I know it will take time. But even after changing everything, I will be going to have to do everything again so I was telling that if you would have merged earlier after final changes it would be good as I hadn't deleted the repo before two days. Anyways, I will be making the changes and making a new PR. And thanks for the guidelines of improving your PR. @rt4914

**Pull request closed without merging**